File: src/augmenters/base_augmenter.py
from abc import ABC, abstractmethod

from src.shockwave_drawer import ShockwaveDrawer


# TODO: fully flesh this out
class TrafficAugmenter(ABC):
    """Abstract base class of all augmenters of the dt-space."""

    @abstractmethod
    def init(self, drawer: ShockwaveDrawer):
        """Initializes an augmenter in the shockwave drawer life cycle. Creates all events the
        augment would generate during the simulation time of the drawer.

        Args:
            simulation_time (float): how long the simulation will last for
            events (SortedList): reference to the event queue
            interfaces (list[Interface]): reference to the list of interfaces
        """
        pass


class CapacityBottleneck(TrafficAugmenter):
    def __init__(self, bottleneck: float):
        self.bottleneck = bottleneck

File: src/augmenters/base_augmenter.ts
import ShockwaveDrawer from "../shockwave_drawer";

/**
 * Base class of anything that would augment/affect traffic flow in the shockwave diagram.
 *
 * @abstract
 * @class TrafficAugmenter
 */
abstract class TrafficAugmenter {
    abstract init(drawer: ShockwaveDrawer): void;
}

/**
 * Specialization of traffic augmenter to augments that affect flow by altering capacity. Currently the only supported type of augment.
 *
 * @export
 * @abstract
 * @class CapacityBottleneck
 * @extends {TrafficAugmenter}
 */
export default abstract class CapacityBottleneck extends TrafficAugmenter {
    public bottleneck: number;

    /**
     * Creates an instance of CapacityBottleneck.
     * 
     * @param {number} bottleneck the bottleneck of this capacity bottleneck
     * @memberof CapacityBottleneck
     */
    constructor(bottleneck: number) {
        super();
        
        this.bottleneck = bottleneck;
    }
}